### PR TITLE
All system contract compatibility and better tests

### PR DESCRIPTION
**Pull request by aaroncox**
Alright… some updates!

I fixed the dot notation issue I mentioned with the eosio contract, which led to a bunch of other things - some of which I also managed to fix. But there’s more :sweat_smile: 

The dot notation thing was that in the system contract, there are tables with . in the name, which when passed in to the factory, was creating object properties with . in them, like powup.state: value and causing errors.

Fix was easy, using JSON.stringify(): https://github.com/wharfkit/contract/pull/52/commits/f6cfcafcd37c4eac101356f21abd93a46945b916

I then went in and rewrote the tests to loop over and test against 4 contracts: eosio , eosio.token, eosio.msig and rewards.gm. All the testing data is setup to process these, but the tests are commented out preventing the failure. To enable them, uncomment here:

https://github.com/wharfkit/contract/blob/6de4f4c913ab960c600fae73156193d1a6ece0cf/test/tests/codegen.ts#L7-L10

and here: 

https://github.com/wharfkit/contract/blob/6de4f4c913ab960c600fae73156193d1a6ece0cf/test/tests/codegen.ts#L25-L32

This is going to cause a number of errors that need to be fixed. There's a bunch of edge cases I don't think we noticed in our limited testing that these additional contracts are now showing. This includes:

## abi.types

The "types" top level field of the ABI needs to be used to resolve mapping issues. These are "mappings" essentially between various types and the eosio contract uses them. At build time we need to read this data and link things properly. Here's the relevant part from the eosio contract abi and an example: 

```json
{
    "version": "eosio::abi/1.2",
    "types": [
        {
            "new_type_name": "block_signing_authority",
            "type": "variant_block_signing_authority_v0"
        }
    ],
    "structs": [
        {
            "name": "block_signing_authority_v0",
            "base": "",
            "fields": [
                {
                    "name": "threshold",
                    "type": "uint32"
                },
                {
                    "name": "keys",
                    "type": "key_weight[]"
                }
            ]
        },
        {
            "name": "regproducer2",
            "base": "",
            "fields": [
                {
                    "name": "producer",
                    "type": "name"
                },
                {
                    "name": "producer_authority",
                    "type": "block_signing_authority"
                },
                {
                    "name": "url",
                    "type": "string"
                },
                {
                    "name": "location",
                    "type": "uint16"
                }
            ]
        },
    ],
    "variants": [
        {
            "name": "variant_block_signing_authority_v0",
            "types": [
                "block_signing_authority_v0"
            ]
        }
    ]
}
``` 

You can see `regproducer2` using `block_signing_authority`. This isn't a struct, it's a top level type, that maps to a variant. Resolving this will lead to the appropriate type...

```
block_signing_authority (regproducer2.producer_authority)
-> variant_block_signing_authority_v0 (from types on the abi)
    -> block_signing_authority_v0 (from variants on the abi)
```

This would mean that on the `regproducer2` action, we should map it like this:

https://github.com/wharfkit/contract/blob/6de4f4c913ab960c600fae73156193d1a6ece0cf/test/data/contracts/mock-eosio.ts#L1117-L1127

On the above link, notice how I already corrected the "mock" version of the code to be what's expected 👍 

## hoisting

Within the type declarations of the generated code we're going to need to somehow do hoisting. 

If you generate the eosio contract, you'll notice that going in order from the ABI the generated code will throw errors. It's because some types are being used before they are defined, like with `ProducerKey` and `ProducerSchedule`. We need the output code to detect this and then order them so these errors don't occur. 

The abi2core library did this, and has a function which might solve this problem:

https://github.com/greymass/abi2core/blob/master/src/toposort.ts

## missing types

Some types aren't resolving properly, you'll notice these if you run against the eosio and eosio.msig contracts. Things like `block_timestamp_type` or `Varuint32`. In the "mock" versions of the contracts I fixed these, but the generator needs to have a mapping for these. 

If you simply run the tests you'll see them!

## values ending with `$`

Unsure what's going on, but there are lots of `$` at the end some values that aren't being removed. You'll see these again if you run the tests.

### CODE REVIEW

File: src/codegen/helpers.ts [hunk 1]
```diff
@@ -1,33 +1,47 @@
...
-    'Name',
-    'TimePoint',
-    'TimePointSec',
-    'UInt128',
-    'UInt16',
-    'UInt32',
-    'UInt64',
-    'UInt8',
-]
+const EOSIO_CORE_CLASSES: string[] = []
```
Comment (dafuga) [inline src/codegen/helpers.ts:7]: We should probably rename this variable too 😄 Maybe  to `ANTELOPE_CLASSES`?

File: src/codegen/interfaces.ts [hunk 2]
```diff
@@ -68,3 +69,21 @@ export function generateActionsNamespace(abi: ABI.Def): ts.ModuleDeclaration {
         ts.NodeFlags.Namespace
     )
 }
+
+function findParamTypeString(typeString: string, namespace: string | null, abi: ABI.Def): string {
```
Comment (dafuga) [inline src/codegen/interfaces.ts:73]: Is this method specific to interfaces? If not, I would put it in the `helper.ts` file.

File: src/codegen/maps.ts [hunk 3]
```diff
@@ -1,12 +1,12 @@
...
+import {ABI} from '@wharfkit/antelope'
 import * as ts from 'typescript'
 import {findAbiType} from './helpers'
 
 export function generateTableMap(abi: ABI.Def): ts.VariableStatement {
     // Map over tables to create the object properties
     const tableProperties = abi.tables.map((table) =>
         ts.factory.createPropertyAssignment(
-            String(table.name),
+            JSON.stringify(table.name),
```
Comment (dafuga) [inline src/codegen/maps.ts:9]: TIL 🤷

File: src/codegen/structs.ts [hunk 4]
```diff
@@ -200,13 +200,17 @@ function findFieldStructTypeString(
...
-
     if (fieldType === 'Symbol') {
         return 'Asset.Symbol'
     }
 
+    if (fieldType === 'Bool') {
+        return 'boolean'
+    }
+
+    if (['String', 'Boolean', 'Number'].includes(fieldType)) {
```
Comment (dafuga) [inline src/codegen/structs.ts:211]: Nice catch here, but if `Boolean` isn't a thing then we should probably remove it from this line.

File: src/codegen/helpers.ts (hunk 1 above)
Comment (aaroncox) [reply src/codegen/helpers.ts:7]: Totally agreed, didn't even think about that 😂

File: src/codegen/interfaces.ts (hunk 2 above)
Comment (aaroncox) [reply src/codegen/interfaces.ts:73]: I think it is exclusive to the interfaces, since it's only used for `ActionParams` as it finds the loose types like `NameType` or `IntType`.

File: src/codegen/maps.ts (hunk 3 above)
Comment (aaroncox) [reply src/codegen/maps.ts:9]: Right? lol....

File: src/codegen/structs.ts (hunk 4 above)
Comment (aaroncox) [reply src/codegen/structs.ts:211]: I'm not sure if it is or isn't a thing, so I didn't want to remove it 😅

File: test/tests/generic.ts [hunk 5]
```diff
@@ -0,0 +1,77 @@
+import {Action, Name} from '@wharfkit/antelope'
```
Comment (dafuga) [inline test/tests/generic.ts:1]: I wonder if we need a `tests/helpers` folder for functions like these 🤔

File: test/tests/table.ts [hunk 6]
```diff
@@ -330,11 +361,20 @@ suite('Table', () => {
...
             assert.instanceOf(row.pervote_bucket, Int64)
         })
+
+        test('should use scope from table call', async function () {
+            const contractKit = new ContractKit({
+                client: makeClient('https://jungle4.greymass.com'),
+            })
+            const contract = await contractKit.load('eosio.token')
+            const result = await contract.table('accounts', 'wharfkittest').get()
+            assert.instanceOf(result.balance, Asset)
```
Comment (dafuga) [inline test/tests/table.ts:376]: Does this test really make sure that the correct scope is used? 🤔 Wouldn't the balance always be an Asset?

**Review by dafuga — APPROVED** [review]
I left you some comments, but it looks great! I think that we can go ahead and merge this whenever you're ready

**Comment by aaroncox** [comment]
@dafuga I didn't want to merge this yet. Instead I figured I'd just hand this branch off to you to work with, since the original comment on this PR still has a bunch of code generation stuff that needs to be addressed to fix the failing tests within.

**Comment by dafuga** [comment]
Sounds good! I'll make a new PR using this as a base branch and we can just merge them both at the same time.
